### PR TITLE
Create vectordb retrieval

### PR DESCRIPTION
**Pull request by bwook00**
- Implement vectordb retrieval using chromadb (the code itself is similar to bm25).
- Add the ability to use vectordb in `retrieval/base.py`.
- push chroma file for test code as well 
(changed the distance metric when creating a chromadb collection to enable 'cosine similarity search'.
 For that db, ingest the corpus.parquet file in the sample project folder)

-> I'll also create a function to ingest vectordb and add it to evaluator.py, then close Feature/#21.

### CODE REVIEW

File: autorag/nodes/retrieval/base.py [hunk 1]
```diff
@@ -28,11 +30,24 @@ def wrapper(
...
-        bm25_path = os.path.join(resources_dir, 'bm25.pkl')
+
         if func.__name__ == "bm25":
+            # check if bm25_path and file exists
+            bm25_path = os.path.join(resources_dir, 'bm25.pkl')
             assert bm25_path is not None, "bm25_path must be specified for using bm25 retrieval."
             assert os.path.exists(bm25_path), f"bm25_path {bm25_path} does not exist. Please ingest first."
-        # TODO: add chroma check for vectordb
+
+        if func.__name__ == "vectordb":
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/base.py:40]: How about using elif?

File: autorag/nodes/retrieval/base.py [hunk 2]
```diff
@@ -28,11 +30,24 @@ def wrapper(
...
+            # check if chroma_path and file exists
+            chroma_path = os.path.join(resources_dir, 'chroma')
+            embedding_model_str = kwargs.pop("embedding_model")
+            chroma_collection_path = os.path.join(chroma_path, embedding_model_str)
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/base.py:44]: delete this unused value.

File: autorag/nodes/retrieval/base.py [hunk 3]
```diff
@@ -28,11 +30,24 @@ def wrapper(
...
+            assert chroma_path is not None, "chroma_path must be specified for using vectordb retrieval."
+            assert os.path.exists(chroma_path), f"chroma_path {chroma_path} does not exist. Please ingest first."
+            # assert chroma_collection_path is not None, \
+                # "chroma_collection_path must be specified for using vectordb retrieval."
+            # assert os.path.exists(chroma_collection_path), \
+                # f"chroma_collection_path {chroma_collection_path} does not exist. Please ingest first."
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/base.py:50]: delete this lines

File: autorag/nodes/retrieval/base.py [hunk 4]
```diff
@@ -41,14 +56,17 @@ def wrapper(
...
 
-        bm25_corpus = load_bm25_corpus(bm25_path)
-
         # run retrieval function
         if func.__name__ == "bm25":
+            bm25_corpus = load_bm25_corpus(bm25_path)
             ids, scores = func(queries=queries, bm25_corpus=bm25_corpus, *args, **kwargs)
+        elif func.__name__ == "vectordb":
+            chroma_collection = load_chroma_collection(db_path=chroma_path, collection_name=embedding_model_str)
+            embedding_model = embedding_models[embedding_model_str]
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/base.py:65]: raise error when there is no embedding_model_str key in embedding_models.

File: autorag/nodes/retrieval/vectordb.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    :param queries: A list of query strings.
+    :param top_k: The number of passages to be retrieved.
+    :param collection: A chroma collection instance that will be used to retrieve passages.
+    :param embedding_model: An embedding model instance that will be used to embed queries.
+
+    :return: The tuple contains a list of passage ids that retrieved from vectordb and a list of its scores.
+    """
+    # embed query
+    embedded_queries = (embedding_model.get_query_embedding(query) for query in queries)
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/vectordb.py:53]: use list(map()) instead of for-loop comprehension.

File: autorag/utils/util.py [hunk 6]
```diff
@@ -9,8 +9,11 @@
 def fetch_contents(corpus_data: pd.DataFrame, ids: List[List[str]]) -> List[List[str]]:
     assert isinstance(ids[0], list), "ids must be a list of list of ids."
     id_df = pd.DataFrame(ids, columns=[f'id_{i}' for i in range(len(ids[0]))])
-    contents_df = id_df.swifter.applymap(
-        lambda x: corpus_data.loc[lambda row: row['doc_id'] == x]['contents'].values[0])
+    try:
+        contents_df = id_df.swifter.applymap(
+            lambda x: corpus_data.loc[lambda row: row['doc_id'] == x]['contents'].values[0])
+    except IndexError:
```
Comment (vkehfdl1) [inline autorag/utils/util.py:15]: add logger.Error for logging error.
`logger.error("doc_id does not exist in corpus_data.")`

File: tests/autorag/nodes/retrieval/test_vectordb.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import os
+import pathlib
+
+import pandas as pd
```
Comment (vkehfdl1) [inline tests/autorag/nodes/retrieval/test_vectordb.py:4]: delete unused import pandas

File: autorag/nodes/retrieval/vectordb.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+import asyncio
+from typing import List, Tuple
+
+import chromadb
```
Comment (vkehfdl1) [inline autorag/nodes/retrieval/vectordb.py:4]: add dependency `chromadb` to requirements.txt

**Review by vkehfdl1 — CHANGES_REQUESTED** [review]

**Review by vkehfdl1 — APPROVED** [review]
Quickly add ingest code and get rid of all chroma resources haha

**Review by Eastsidegunn — APPROVED** [review]
LGTM

**Review by Eastsidegunn — APPROVED** [review]
LGTM

**Review by Eastsidegunn — DISMISSED** [review]
LGTM

**Review by vkehfdl1 — APPROVED** [review]
LGTM

**Review by Eastsidegunn — APPROVED** [review]

**Pull request merged**